### ISSUE TITLE
Crash on notifications background fetch

### ISSUE THREAD

**Issue by koke** [wordpress-mobile/WordPress-iOS#7216]
It first seemed like a new crash, but Fabric says it's been around since 7.0

```
-[ContextManager saveContextAndWait:] (ContextManager.m:145)
specialized NotificationSyncMediator.(markAsRead(Notification, completion : (Error?) -> ()?) -> ()).(closure #1) (NotificationSyncMediator.swift)
partial apply for NotificationSyncMediator.(markAsRead(Notification, completion : (Error?) -> ()?) -> ()).(closure #1) (NotificationSyncMediator.swift)
partial apply for WordPressComRestApi.(GET(String, parameters : [String : AnyObject]?, success : (AnyObject, HTTPURLResponse?) -> (), failure : (NSError, HTTPURLResponse?) -> ()) -> Progress?).(closure #3) (WordPressComRestApi.swift)
__116-[AFHTTPSessionManager dataTaskWithHTTPMethod:URLString:parameters:uploadProgress:downloadProgress:success:failure:]_block_invoke.80 (AFHTTPSessionManager.m:290)
__72-[AFURLSessionManagerTaskDelegate URLSession:task:didCompleteWithError:]_block_invoke (AFURLSessionManager.m:282)
```

And the actual error:

`Unresolved core data error
Error Domain=NSCocoaErrorDomain Code=132001 "(null)" UserInfo={message=attempt to recursively call -save: on the context aborted`

Every time this happens, right before the core data error there's this in the log:

` Error marking note as read: Error Domain=NSURLErrorDomain Code=-1001 "The request timed out." UserInfo={NSUnderlyingError=0x17485cbc0 {Error Domain=kCFErrorDomainCFNetwork Code=-1001 "(null)" UserInfo={_kCFStreamErrorCodeKey=-2102, _kCFStreamErrorDomainKey=4}}, NSErrorFailingURLStringKey=https://public-api.wordpress.com/rest/v1.1/notifications/read?locale=en, NSErrorFailingURLKey=https://public-api.wordpress.com/rest/v1.1/notifications/read?locale=en, _kCFStreamErrorDomainKey=4, _kCFStreamErrorCodeKey=-2102, NSLocalizedDescription=The request timed out.}`

**Comment by koke** [comment]
There are several odd things with this one:

- Whenever there's a push notification, we call `NotificationSyncMediator.sync`, even if there's a sync in progress.
- This happens with the app in the background, `NotificationDetailsViewController` should not be marking anything as read.

I think the scenario to reproduce this is a very specific race condition (which I don't fully understand yet), but it'd still be good to address the underlying issues.

**Comment by koke** [comment]
I can reproduce if I use [charles](http://charlesproxy.com) and force the mark as read request to fail:

![screen shot 2017-05-23 at 17 10 23](https://cloud.githubusercontent.com/assets/8739/26361691/029b09ea-3fdc-11e7-85d6-ab04737ec812.png)

Then:
- Open a post like notification
- Put the app in the background
- Like that post from another account
- Receive a push notification for it
- 💥

**Comment by koke** [comment]
Actually, there's a much easier (and more terrible) way to reproduce this:
- Put your phone in airplane mode
- Tap on any unread notification
- 💥

**Comment by bummytime** [comment]
This was fixed in #7226.
